src/engines/alsadevicefinder.cpp

```diff
@@ -0,0 +1,110 @@
+/* This file is part of Clementine.
+   Copyright 2017, Jonas Kvinge <jonas@jkvinge.net>
+
+   Clementine is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   Clementine is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with Clementine.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <alsa/asoundlib.h>
+#include <errno.h>
+#include <stdio.h>
+
+#include <QList>
+#include <QString>
+#include <QVariant>
+
+#include <core/logging.h>
+
+#include "alsadevicefinder.h"
+#include "devicefinder.h"
+
+AlsaDeviceFinder::AlsaDeviceFinder() : DeviceFinder("alsasink") {}
+
+QList<DeviceFinder::Device> AlsaDeviceFinder::ListDevices() {
+  QList<Device> ret;
+
+  int result = -1;
+  int card = -1;
+  int dev = -1;
+
+  snd_ctl_card_info_t* cardinfo;
+  snd_pcm_info_t* pcminfo;
+  snd_ctl_t* handle;
+
+  snd_ctl_card_info_alloca(&cardinfo);
+  snd_pcm_info_alloca(&pcminfo);
+
+  snd_pcm_stream_name(SND_PCM_STREAM_PLAYBACK);
+
+  while (true) {
+    result = snd_card_next(&card);
+    if (result < 0) {
+      qLog(Error) << "Unable to get soundcard:" << snd_strerror(result);
+      break;
+    }
+    if (card < 0) break;
+
+    char name[32];
+    sprintf(name, "hw:%d", card);
+
+    result = snd_ctl_open(&handle, name, 0);
+    if (result < 0) {
+      qLog(Error) << "Unable to open soundcard" << card << ":"
+                  << snd_strerror(result);
+      continue;
+    }
+    result = snd_ctl_card_info(handle, cardinfo);
+    if (result < 0) {
+      qLog(Error) << "Control hardware failure for card" << card << ":"
+                  << snd_strerror(result);
+      snd_ctl_close(handle);
```

Comment:
You could use boost scope exit to make these handle closures safer, something like:
```c++
BOOST_SCOPE_EXIT(&handle) {
  snd_ctl_close(handle);
} BOOST_SCOPE_EXIT_END
```